subsys/net/conn_mgr/conn_mgr_monitor.c

```diff
@@ -144,6 +727,8 @@ static void conn_mgr_mon_handle_update(void)
 		} else if (original_ready_count == 0) {
 			/* We just gained connectivity */
 			net_mgmt_event_notify(NET_EVENT_L4_CONNECTED, last_iface_up);
+
+			trigger_online_connectivity_check();
```

Comment:
I am not sure if we need to fire multiple online events, I mean from the system point of view it should be enough to have online connectivity once and not do it for each interface. If there are multiple network interfaces, normally it is enough to have for example Internet connectivity via one network interface.

Reply:
> and not do it for each interface

I definitely do not think you should fire online for each interface -- note rule 2:

> The online notification cannot fire again until all ifaces have gone offline

But I think internally Online-ness needs to be tracked for each interface

Consider the following edge case:

1. if1 and if2 become ready together, but if2 cannot reach online state
2. if1 reaches online state

Now, the ONLINE event is triggered

Now, suppose `if1` goes oper-down. In theory, our only source of online-ness has been lost.

If your trickle feature is enabled, it will probably detect this eventually, but if we tracked online-ness on a per-iface level, we could know immediately just from the oper-down state that we should now fire the OFFLINE event.

Furthermore, suppose your trickle feature is not enabled: Then if if2 never goes oper-down, then the OFFLINE event will never be fired at all.

Reply:
I will shortly send a new version for review, it will not have changes in this front. I need to think more what/how to do with this.

Reply:
After looking the code and the discussion above, I am not fully following the logic here. I placed the online/offline checks to the same place where `L4_CONNECTED/DISCONNECTED` events are generated. This means that the system cannot get to online state unless it has first entered `L4_CONNECTED` state.
How is the online/offline handling different from connected/disconnected handling?

Reply:
> How is the online/offline handling different from connected/disconnected handling?

There isn't. The Connected event fires when the ***first*** iface goes oper-up. The Disconnected event fires when the ***last*** iface goes oper-down.

Similarly, the Online event should fire when the ***first*** iface goes ***online***, and the Offline event should fire when the ***last*** iface goes ***offline***.

Currently, you have it so that the Offline event can only fire when all ifaces are oper-down. So the only available ***online*** iface might go ***oper-down***, but this can end up ***not*** firing the Offline event if there is another iface that is still oper-up (but not online).

Reply:
> Similarly, the Online event should fire when the _**first**_ iface goes _**online**_, and the Offline event should fire when the _**last**_ iface goes _**offline**_. (and an iface should be considered offline if it is oper-down/unready)
> 
> Currently, you have it so that the Offline event can only fire when all ifaces are oper-down. So the only available _**online**_ iface might go _**oper-down**_, but this can end up _**not**_ firing the Offline event if there is another iface that is still oper-up (but not online).

What you are describing is not how online/offline should work. As the online/offline status is tied to L4_connected/disconnected, if we get a L4_disconnected, then there cannot be any online interfaces, so offline event is triggered. Note that online/offline is just an extra property to L4 connected/disconnected, as online/offline status just does extra checks whether there is a connection is to some predefined host.

I will need to add support for checking if we were L4_connected, and get another interface L4_connected and we were offline, then we need to do online check. But if we are already online, there should be no need to do any additional online checks as one online interface should be enough.

Reply:
> if we get a L4_disconnected, then there cannot be any online interfaces

That is true, but I am trying to point out that if you do things this way, it is possible for the only online interface to become unready (L4-down), but for L4_disconnected to not be fired

Reply:
> Specifically, this happens if there is another separate iface which is ready (iface oper-up), but never passes online checks.
> 
> If that limitation is intentional, then go for it, but it seems like it is not?

Not intentional, and yes we should check what you describe. I will try to address this case next.

> IMO, if the only online iface goes L4-oper-down, it should fire L4_offline, since after that happens there are no longer any online ifaces

Yes, definitely. And I think that happens already in `conn_mgr_mon_handle_update()` which check if there are any ready interfaces and triggers L4_disconnected and offline event.